### ISSUE TITLE
useActionData can't return data from different action

### ISSUE THREAD

**Issue by jaschaio** [remix-run/remix#3564]
### What version of Remix are you using?

1.6.1

### Steps to Reproduce

Sample repository: https://github.com/jaschaio/remix-use-action-data

Simply run `npx create-remix@latest`

The only files added or modified are:

* `/app/context/modal.js`
* `/app/components/modal.js`
* `/app/routes/index.jsx`
* `/app/root.jsx`

The `/app/routes/index.jsx` route simply defines an `action` and renders a button to open a modal component:

```jsx
import { useModal } from '~/context/modal';

export const action = () => {

    return 'done';

};

export default function Index() {

    const [ modal, setModal ] = useModal();

    return (
        <button onClick={ () => setModal( true ) }>Open Modal</button>
    );

}
```

The modal context is defined in `/app/context/modal.js` and a simple `useState` on/off switch:

```js
import { createContext, useContext, useState } from 'react';

const Context = createContext();

export const ProvideModal = ( { children } ) => {

    return <Context.Provider value={ useState( false ) } >{ children }</Context.Provider>

};

export const useModal = () => {

    return useContext( Context );

};
```

The modal component `/app/components/modal.js` renders a modal which contains a `Form`:

```jsx
import { useRef } from 'react';
import { Form, useActionData } from '@remix-run/react';

import { useModal } from '~/context/modal';

const Modal = () => {

    const ref = useRef();

    const data = useActionData( ref );

    console.log( { useActionData: data } );

    const [ modal, setModal ] = useModal();

    if ( ! modal )
        return null;

    return (
        <dialog open>
            <Form method="post" action="/?index" ref={ ref }>
                <input type="text" placeholder="Write something…" />
                <button type="submit">Submit</button>
            </Form>
            <button onClick={ () => setModal( false ) }>Close Modal</button>
        </dialog>
    );

};

export default Modal;
```

The `/app/root.jsx` component simply makes sure to render the modal when it's open and provide the necessary context:

```diff
import {
  Links,
  LiveReload,
  Meta,
  Outlet,
  Scripts,
  ScrollRestoration,
} from "@remix-run/react";

+ import { ProvideModal } from '~/context/modal';

+ import Modal from '~/components/Modal';

export const meta = () => ({
  charset: "utf-8",
  title: "New Remix App",
  viewport: "width=device-width,initial-scale=1",
});

export default function App() {

  return (
    <html lang="en">
      <head>
        <Meta />
        <Links />
      </head>
      <body>
+        <ProvideModal>
            <Outlet />
+            <Modal />
+        </ProvideModal>
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}
```

You can run the app via cloning the directory `npm i` and `npm run dev`

### Expected Behavior

When a form with a different action is submitted, it should be possible to either pass a reference or an action url to `useActionData`. I believe @ryanflorence actually discussed this in #151 but passing in a `ref` to `useActionData` does not work anymore.

### Actual Behavior

`useActionData` will always return undefined if it's called from a component that is not placed in the original route file

**Comment by chaance** [comment]
I don't believe this is a bug. If you check the docs for `useActionData`:

> This hook returns the JSON parsed data from your route action. **It returns `undefined` if there hasn't been a submission at the current location yet.**

Consider the context hierarchy here:

- Each route renders a `<Route />` component, which is what provides context for actions/loaders via `RemixRouteContext`
- Child routes will have access to the context of their parents, but parents will not have access to context of their children
- `root.jsx` is a parent of `routes/index.jsx`, so it will not have access to the action data provided via the context from the index route
- `<Modal>` is rendered in `root.jsx`, so it makes sense why it can't access the action data

In order for your modal to have access to the action data, it would need to either be rendered within the route hierarchy of the action, or you'd need to use `useFetcher().Form` which is specifically for fetching data at other routes.

```jsx
import { useRef, useState, useEffect } from "react";
import { Form, useActionData, useFetcher, useNavigate } from "@remix-run/react";

import { useModal } from "~/context/modal";

const Modal = () => {
  let [modal, setModal] = useModal();
  let fetcher = useFetcher();
  let navigate = useNavigate();

  // you will see the data here after submitting
  console.log({ actionData: fetcher.data });

  // A regular `Form` does a navigation to the action route, but a fetcher
  // `Form` does not. If you do actually want to navigate to the route after
  // submission you can do so imperatively by reading the fetcher type
  useEffect(() => {
    if (fetcher.type === "done") {
      navigate("/?index");
    }
  }, [fetcher.type]);

  if (!modal) {
    return null;
  }

  return (
    <dialog open>
      <fetcher.Form method="post" action="/?index">
        <input type="text" placeholder="Write something…" />
        <button type="submit">Submit</button>
      </fetcher.Form>
      <button onClick={() => setModal(false)}>Close Modal</button>
    </dialog>
  );
};
```

**Comment by jaschaio** [comment]
I agree, not a bug. But it would be a really nice feature, as it goes against developer expectation and adds to the problem of route files getting very long and complicated in the worst case.

**Comment by chaance** [comment]
I'd disagree with you there. One-way data flow, this is just how context works. I wouldn't expect a parent route to get action data from all of its children. In fact, I'd say it would be actively bad to expect that it would, seeing as how different children could easily respond with conflicting data.

**Comment by jaschaio** [comment]
I was typing a response but than took another look at your suggested solution:

> it would need to either be rendered within the route hierarchy of the action, or you'd need to use `useFetcher().Form` which is specifically for fetching data at other routes

I missed the `useFetcher().Form` part when I first read it and thought you were only suggesting `useFetcher()` and the `load` or `submit` methods which obviously would break progressive enhancement. I wasn't even aware of `useFetcher().Form`.

Thanks for the suggestion, I will rework my example and see if that solves the problem, as rendering a generic Modal Component way down the route hierarchy doesn't make sense as I would have to remember to include it everywhere I want to render a Modal.

**Comment by ludu12** [comment]
@jaschaio did you find any solution to this? My use case is very similar where I want to make a generic modal component. I was sort of hoping that using

```js
const Inputs = () => {
    const actionData = useActionData();
    return <>...</>
}

 <fetcher.Form action={'/action'} method="post">
    <Inputs/>
</fetcher.Form
```

Would just provide the appropriate context or something, but that doesn't seem to be the case. Still new to remix and maybe my mental model isn't quite right yet!
